sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,530 @@
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be opened for writing
+     * @param context the IO context
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        if (name.contains("segments_") || name.endsWith(".si")) return super.createOutput(name, context);
+        ensureOpen();
+        OutputStream fos = Files.newOutputStream(directory.resolve(name), StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
+        Cipher cipher = CipherFactory.getCipher(provider);
+        SecureRandom random = Randomness.createSecure();
+        byte[] iv = new byte[CipherFactory.IV_ARRAY_LENGTH];
+        random.nextBytes(iv);
+        if (dataKey == null) throw new RuntimeException("dataKey is null!");
+        CipherFactory.initCipher(cipher, this, Optional.of(iv), Cipher.ENCRYPT_MODE, 0);
+        ivMap.put(getDirectory() + "/" + name, Base64.getEncoder().encodeToString(iv));
+        return new CryptoIndexOutput(name, fos, cipher);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param prefix the desired temporary file prefix
+     * @param suffix the desired temporary file suffix
+     * @param context the IO context
+     */
+    @Override
+    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
+        if (prefix.contains("segments_") || prefix.endsWith(".si")) return super.createTempOutput(prefix, suffix, context);
+        ensureOpen();
+        String name;
+        while (true) {
+            name = getTempFileName(prefix, suffix, nextTempFileCounter.getAndIncrement());
+            OutputStream fos = Files.newOutputStream(directory.resolve(name), StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
+            Cipher cipher = CipherFactory.getCipher(provider);
+            SecureRandom random = Randomness.createSecure();
+            byte[] iv = new byte[CipherFactory.IV_ARRAY_LENGTH];
+            random.nextBytes(iv);
+            CipherFactory.initCipher(cipher, this, Optional.of(iv), Cipher.ENCRYPT_MODE, 0);
+            ivMap.put(getDirectory() + "/" + name, Base64.getEncoder().encodeToString(iv));
+            return new CryptoIndexOutput(name, fos, cipher);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void close() throws IOException {
+        try {
+            deleteFile("ivMap");
+        } catch (java.nio.file.NoSuchFileException fnfe) {
+
+        }
+        IndexOutput out = super.createOutput("ivMap", new IOContext());
+        out.writeMapOfStrings(ivMap);
+        out.close();
+        isOpen = false;
+        deletePendingFiles();
+        dataKey = null;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be deleted
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
```

Comment:
I am not sure how this will work. I am sure you know better than me about the NiofsDIrectory you are extending here, but my understanding is that Lucene does not physically delete indices even when you trigger a delete. Instead it works similar to system memory where we only delete on overwrite or an explicit block clear. I don't know how that interacts here but if you are needing to assume true deletion at this point then I think we will see some trouble here.

Reply:
this method corresponds to deleting a file within an index (directory) and not the index itself. There is a test for this in `CryptoDirectoryTests::testDeleteFile` which passes.